### ISSUE TITLE
Extending knitr. Add package to preamble

### ISSUE THREAD

**Issue by zeehio** [yihui/knitr#1184]
Hi,

First of all thanks for your work.

I am working on a package for data frame visualization [condformat](http://www.sergioller.com/2016-03-10-condformat-release.md). Summarizing, it allows to add conditional formatting rules to a data frame, and then it can convert the data frame to HTML or LaTeX.

The generated latex code relies on `\usepackage[table]{xcolor}`, and I would like to be able to add it to the preamble of the documents that use it. To give a simple example:

1. Install the package:

        devtools::install_github("zeehio/condformat")

2. Save this R-code on a `my_file.R`

        #'---
        #' title: "Titre"
        #' date: Fecha
        #' header-includes:
        #'  - \usepackage[table]{xcolor}
        #' output:
        #'  pdf_document:
        #'    keep_tex: TRUE
        #'---
        data(iris)
        library(condformat)
        mydataset <- iris[c(1:5,70:75, 120:125),]
        cosa <- condformat(mydataset) +
          rule_fill_discrete(Species) +
          rule_fill_discrete(Sepal.Width, Sepal.Length,
                             expression = Sepal.Width > Sepal.Length - 2.25,
                             colours = c("TRUE" = "#7D00FF")) +
          rule_fill_gradient2(Petal.Length)
        
        cosa
        
    

3. Open `my_file.R` and click on RStudio File/Compile Notebook.

As the `condformat` package maintainer, I would like to add  `\usepackage[table]{xcolor}` to the LaTeX preamble, however `knitr::set_header` does not seem to allow me to define my own packages (only `tikz`, `highlight` and few others are available).

Is there a way for me to add the LaTeX package in the preamble without writing the `header-includes` on each file? (Ideally `condformat` users would not need to know about LaTeX)

Thanks

**Comment by yihui** [comment]
`knitr::set_header` won't work for R Markdown v2. It is for Rnw and Rhtml documents only.

Ideally, we should have a similar thing for LaTeX output just like `htmltools::htmlDependency()` for HTML output, but we do not have that, so the only way to insert code into the LaTeX preamble is the `header-includes` variable. Actually that is not the optimal way, since it will also be inserted into the HTML output. You should use the `includes` option of `pdf_document` instead: http://rmarkdown.rstudio.com/pdf_document_format.html#includes (which means you have to put the code in a separate file).

Although I see the point, I guess it is unlikely to be implemented in rmarkdown (at least not soon).

**Comment by zeehio** [comment]
I was trying to do the dirty trick:

     env <- environment(rmarkdown::render)
     env$metadata$output$pdf_document$includes$in_header <- "xcolor.tex"

But I can't overwrite the `rmarkdown::metadata` variable (or the variable is used to write the tex header before I can modify it)...

In any case, thank you for your support. If any dirty-not-future-proof hack happens to go through your mind I will happily try it...

Thinking about providing a proper solution to this, on the one hand I could contribute a pull request that updates the `output_format` variable after calling `knit::knit` in [rmarkdown::render](https://github.com/rstudio/rmarkdown/blob/693b0b1ce2b679c411c4ad4553612e2c943d90f6/R/render.R#L294)... do you think it could be welcome? If you have any idea of what is the right place to implement a solution I would do my best to provide a good patch.

On the other hand on the knitr side I could provide a patch to have `knitr::latexDependency(pkg_name, pkg_options)` that asks for a latex package name and package options and puts that into a list. With a `knitr:::renderLatexDependencies()` that collates list in a latex valid header. Then inserting that generated header in the knitr output header should not be hard... your advice on this would be very welcome, if you feel this is a proper solution.

**Comment by yihui** [comment]
`rmarkdown::metadata` is supposed to be read-only.

If we were to solve this problem anyway, I'd prefer the `latexDependency()` approach. Please feel free to work on that. I guess it is probably not super hard, but you will have to learn a couple of things: [`knitr::knit_print()`](https://cran.rstudio.com/web/packages/knitr/vignettes/knit_print.html) (how to emit metadata), `htmltools::htmlDependency()` (you only need to know the idea), and how **rmarkdown** resolves the knitr metadata (HTML dependencies) into a temporary file and inserts that into the HTML head.

**Comment by zeehio** [comment]
Thanks @yihui. I have contributed a [pull request](https://github.com/rstudio/rmarkdown/pull/647) in the [rmarkdown](rstudio/rmarkdown) page. I hope to have understood properly your suggestions.

Comments and code reviews would be very welcome.

If you feel this is better discussed at the pull request thread I opened, feel free to close this issue, as (as far as I see it) it does not fit into the `knitr` package.

Thanks again for your time
